linkerd/trace-context/src/propagation.rs

```diff
@@ -42,16 +47,111 @@ impl TraceContext {
+        Propagation::B3Grpc => increment_grpc_span_id(request, context),
+        Propagation::B3Http => increment_http_span_id(request),
+        Propagation::W3CHttp => {
+            increment_w3c_http_span_id(request, context.trace_id.clone(), context.flags.clone())
+        }
+    }
+}
+
+fn unpack_w3c_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
+    get_header_str(request, W3C_HTTP_TRACEPARENT).and_then(parse_w3c_context)
+}
+
+fn parse_w3c_context(header_value: &str) -> Option<TraceContext> {
+    let rest = match header_value.split_once(W3C_HEADER_VALUE_SEPARATOR) {
+        Some((version, rest)) => {
+            if version != W3C_HEADER_VERSION {
+                warn!(
+                    "Tracecontext header {} contains invalid version: {}",
+                    header_value, version
+                );
+                return None;
+            }
+            rest
+        }
+        None => {
+            warn!("Tracecontext header {} is invalid", header_value);
+            return None;
+        }
+    };
+
+    let (trace_id, rest) = if let Some((id, rest)) = parse_w3c_header_value(rest, 16) {
+        (id, rest)
+    } else {
+        warn!("Tracecontext header {} contains invalid id", header_value);
+        return None;
+    };
+
+    let (parent_id, rest) = if let Some((id, rest)) = parse_w3c_header_value(rest, 8) {
+        (id, rest)
+    } else {
+        warn!("Tracecontext header {} contains invalid id", header_value);
+        return None;
+    };
+
+    let flags = match hex::decode(rest) {
+        // If valid hex, take final bit and AND with 1. W3C only uses one bit
+        // for flags in version 00, and the bit is used to control sampling
+        Ok(decoded) => Flags(decoded[0] & 0x1),
+        // If invalid hex, do not sample trace
+        Err(e) => {
+            warn!("Failed to decode flags for header {}: {}", header_value, e);
+            Flags(0)
+        }
+    };
+
+    Some(TraceContext {
+        propagation: Propagation::W3CHttp,
+        trace_id,
+        parent_id,
+        flags,
+    })
+}
+
+fn parse_w3c_header_value(next_header: &str, pad_to: usize) -> Option<(Id, &str)> {
+    next_header
+        .split_once(W3C_HEADER_VALUE_SEPARATOR)
+        .filter(|(id, _rest)| !id.chars().all(|c| c == '0'))
+        .and_then(|(id, rest)| decode_id_with_padding(id, pad_to).zip(Some(rest)))
+}
+
+fn increment_w3c_http_span_id<B>(request: &mut http::Request<B>, trace_id: Id, flags: Flags) -> Id {
+    let span_id = Id::new_span_id(&mut thread_rng());
+
+    trace!("incremented span id: {}", span_id);
+
+    let new_header = {
+        let mut buf = String::with_capacity(60);
+        buf.push_str(W3C_HEADER_VERSION);
+        buf.push_str(W3C_HEADER_VALUE_SEPARATOR);
+        buf.push_str(&hex::encode(trace_id));
+        buf.push_str(W3C_HEADER_VALUE_SEPARATOR);
+        buf.push_str(&hex::encode(span_id.as_ref()));
+        buf.push_str(W3C_HEADER_VALUE_SEPARATOR);
```

Comment:
Similarly, I think this is fine.

```suggestion
        buf.push('-');
        buf.push_str(&hex::encode(trace_id));
        buf.push('-');
        buf.push_str(&hex::encode(span_id.as_ref()));
        buf.push('-');
```